backend-db/Rhyolite/Backend/TaskWorker.hs

```diff
@@ -8,19 +10,36 @@ import Rhyolite.Backend.Schema.Task
 import Control.Concurrent
 import Control.Concurrent.Async
 import Control.Concurrent.Thread.Delay
-import Control.Exception.Lifted (bracket)
+import Control.Exception.Lifted (bracket, catch, throw, Exception, SomeException(..))
 import Control.Monad
 import Control.Monad.IO.Class
 import Control.Monad.Logger
 import Control.Monad.Trans.Control
 import Data.Functor.Identity
 import Data.Pool
 import Data.Text (Text)
+import qualified Data.Text as T
 import Data.Time
 import Database.Groundhog.Core
 import Database.Groundhog.Expression
 import Database.Groundhog.Postgresql
+import qualified Database.PostgreSQL.Simple.Types as PG
 import Rhyolite.Backend.DB
+import Rhyolite.Backend.DB.PsqlSimple (PostgresRaw, executeQ)
+
+withSavepoint
+  :: (PostgresRaw m, MonadBaseControl IO m, Exception e)
+  => String -> m a -> m (Either e a)
+withSavepoint name action = do
+  let savePt = PG.Identifier $ T.pack name
+  [executeQ|SAVEPOINT ?savePt|]
+  result <- catch (Right <$> action) $ \e -> return (Left e)
```

Comment:
Shouldn't this be using `bracket`? An exception can leave a savepoint in place. Also it seems like `name` should be randomly generated instead.